// ix/Ix.dom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncIterableReadableByteStream = exports.AsyncIterableReadableStream = exports.fromDOMStream = exports.AsyncIterable = exports.Iterable = exports.AsyncSink = exports.OrderedAsyncIterableBase = exports.OrderedAsyncIterable = exports.OrderedIterableBase = exports.OrderedIterable = exports.symbolObservable = exports.default = void 0;
const tslib_1 = require("tslib");
// Manually re-export because closure-compiler doesn't support `export * from X` syntax yet
tslib_1.__exportStar(require("./Ix"), exports);
var Ix_1 = require("./Ix");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return Ix_1.default; } });
var Ix_2 = require("./Ix");
Object.defineProperty(exports, "symbolObservable", { enumerable: true, get: function () { return Ix_2.symbolObservable; } });
var Ix_3 = require("./Ix");
Object.defineProperty(exports, "OrderedIterable", { enumerable: true, get: function () { return Ix_3.OrderedIterable; } });
var Ix_4 = require("./Ix");
Object.defineProperty(exports, "OrderedIterableBase", { enumerable: true, get: function () { return Ix_4.OrderedIterableBase; } });
var Ix_5 = require("./Ix");
Object.defineProperty(exports, "OrderedAsyncIterable", { enumerable: true, get: function () { return Ix_5.OrderedAsyncIterable; } });
var Ix_6 = require("./Ix");
Object.defineProperty(exports, "OrderedAsyncIterableBase", { enumerable: true, get: function () { return Ix_6.OrderedAsyncIterableBase; } });
var Ix_7 = require("./Ix");
Object.defineProperty(exports, "AsyncSink", { enumerable: true, get: function () { return Ix_7.AsyncSink; } });
Object.defineProperty(exports, "Iterable", { enumerable: true, get: function () { return Ix_7.Iterable; } });
Object.defineProperty(exports, "AsyncIterable", { enumerable: true, get: function () { return Ix_7.AsyncIterable; } });
var fromdomstream_1 = require("./asynciterable/fromdomstream");
Object.defineProperty(exports, "fromDOMStream", { enumerable: true, get: function () { return fromdomstream_1.fromDOMStream; } });
Object.defineProperty(exports, "AsyncIterableReadableStream", { enumerable: true, get: function () { return fromdomstream_1.AsyncIterableReadableStream; } });
Object.defineProperty(exports, "AsyncIterableReadableByteStream", { enumerable: true, get: function () { return fromdomstream_1.AsyncIterableReadableByteStream; } });
