// wechaty-puppet/src/mods/payloads.ts
import type {
  ContactPayload,
  EventDirtyPayload,
  EventDongPayload,
  EventErrorPayload,
  EventFriendshipPayload,
  EventHeartbeatPayload,
  EventLoginPayload,
  EventLogoutPayload,
  EventMessagePayload,
  EventPostPayload,
  EventReadyPayload,
  EventResetPayload,
  EventRoomInvitePayload,
  EventRoomJoinPayload,
  EventRoomLeavePayload,
  EventRoomTopicPayload,
  EventScanPayload,
  FriendshipPayload,
  FriendshipPayloadConfirm,
  FriendshipPayloadReceive,
  FriendshipPayloadVerify,
  LocationPayload,
  MessagePayload,
  MessagePayloadBase,
  MessagePayloadRoom,
  MessagePayloadTo,
  MiniProgramPayload,
  PostPayload,
  PostPayloadClient,
  PostPayloadServer,
  RoomInvitationPayload,
  RoomMemberPayload,
  RoomPayload,
  TapPayload,
  UrlLinkPayload,
  SayablePayload,
}                           from '../schemas/mod.js'
import {
  sayablePayloads,
  isPostPayloadServer,
  isPostPayloadClient,
}                           from '../schemas/mod.js'

export type {
  ContactPayload            as Contact,
  EventDirtyPayload         as EventDirty,
  EventDongPayload          as EventDong,
  EventErrorPayload         as EventError,
  EventFriendshipPayload    as EventFriendship,
  EventHeartbeatPayload     as EventHeartbeat,
  EventLoginPayload         as EventLogin,
  EventLogoutPayload        as EventLogout,
  EventMessagePayload       as EventMessage,
  EventPostPayload          as EventPost,
  EventReadyPayload         as EventReady,
  EventResetPayload         as EventReset,
  EventRoomInvitePayload    as EventRoomInvite,
  EventRoomJoinPayload      as EventRoomJoin,
  EventRoomLeavePayload     as EventRoomLeave,
  EventRoomTopicPayload     as EventRoomTopic,
  EventScanPayload          as EventScan,
  FriendshipPayload         as Friendship,
  FriendshipPayloadConfirm  as FriendshipConfirm,
  FriendshipPayloadReceive  as FriendshipReceive,
  FriendshipPayloadVerify   as FriendshipVerify,
  LocationPayload           as Location,
  MessagePayload            as Message,
  MessagePayloadBase        as MessageBase,
  MessagePayloadRoom        as MessageRoom,
  MessagePayloadTo          as MessageTo,
  MiniProgramPayload        as MiniProgram,
  PostPayload               as Post,
  PostPayloadClient         as PostClient,
  PostPayloadServer         as PostServer,
  RoomInvitationPayload     as RoomInvitation,
  RoomMemberPayload         as RoomMember,
  RoomPayload               as Room,
  SayablePayload            as Sayable,
  TapPayload                as Tap,
  UrlLinkPayload            as UrlLink,
}
export {
  sayablePayloads as sayable, // Sayable payload creators
  isPostPayloadServer as isPostServer,
  isPostPayloadClient as isPostClient,
}
