// wechaty-token/src/grpc-js.ts
import type { ServiceConfig }     from '@grpc/grpc-js/build/src/service-config.js'
import type { StatusObject }      from '@grpc/grpc-js/build/src/call-stream.js'
import type {
  TcpSubchannelAddress,
}                                 from '@grpc/grpc-js/build/src/subchannel-address.js'
import {
  parseUri,
  uriToString,
  type GrpcUri,
}                                 from '@grpc/grpc-js/build/src/uri-parser.js'
import * as resolverManager       from '@grpc/grpc-js/build/src/resolver.js'
import type { ChannelOptions }    from '@grpc/grpc-js/build/src/channel-options.js'
import { Metadata }               from '@grpc/grpc-js/build/src/metadata.js'
import { BackoffTimeout }         from '@grpc/grpc-js/build/src/backoff-timeout.js'

import {
  status as GrpcStatus,
}                                 from '@grpc/grpc-js'

export {
  BackoffTimeout,
  ChannelOptions,
  GrpcStatus,
  type GrpcUri,
  Metadata,
  parseUri,
  resolverManager,
  ServiceConfig,
  StatusObject,
  TcpSubchannelAddress,
  uriToString,
}
