// xstate/es/index.js
import { assign as assign$1, send as send$1, sendParent as sendParent$1, sendUpdate as sendUpdate$1, forwardTo as forwardTo$1, doneInvoke as doneInvoke$1 } from './actions.js';
import * as actions from './actions.js';
export { actions };
export { toActorRef } from './Actor.js';
export { Interpreter, InterpreterStatus, interpret, spawn } from './interpreter.js';
export { Machine, createMachine } from './Machine.js';
export { mapState } from './mapState.js';
export { matchState } from './match.js';
export { createSchema, t } from './schema.js';
export { State } from './State.js';
export { StateNode } from './StateNode.js';
export { spawnBehavior } from './behaviors.js';
export { ActionTypes, SpecialTargets } from './types.js';
export { matchesState, toEventObject, toObserver, toSCXMLEvent } from './utils.js';

var assign = assign$1,
    send = send$1,
    sendParent = sendParent$1,
    sendUpdate = sendUpdate$1,
    forwardTo = forwardTo$1,
    doneInvoke = doneInvoke$1;

export { assign, doneInvoke, forwardTo, send, sendParent, sendUpdate };
